### PR TITLE
Non-transferable items

### LINKED ISSUE #141: Implement non-transferable items
[RMRK2 spec](https://github.com/rmrk-team/rmrk-spec/blob/master/standards/rmrk2.0.0/entities/nft.md#nft) contains `transferable` property currently missing in pallets. ( Needed for ie Kanaria gems and all the "soulbound" mechanics. )

In Kusama’s implementation we also have a modifier for temporary transferability, so that you can say, this becomes non transferable after X blocks, or it is transferable for only X blocks, but to not complicate things just the basic boolean would do.

### PR DESCRIPTION
**Pull request by ilionic**
Fixes #141

### CODE REVIEW

File: pallets/rmrk-core/src/functions.rs [hunk 1]
```diff
@@ -369,6 +373,10 @@ where
 
 		// TODO: Check NFT lock status
 
+		// Check NFT is transferable
```
Comment (HashWarlock) [inline pallets/rmrk-core/src/functions.rs:376]: We could make this an internal public function returns `DispatchResult` so downstream can call a function like `is_transferable(&nft)`

Comment (ilionic) [reply pallets/rmrk-core/src/functions.rs:376]: @HashWarlock Extracted function but passing actual transferable `bool` instead of nft reference. Wdyt?

Comment (bmacer) [reply pallets/rmrk-core/src/functions.rs:376]: if that's the case, just change 377 to
```
		Self::check_is_transferable(&sending_nft)?;
```

and 697-701 to
```
	// Check NFT is transferable
	pub fn check_is_transferable(nft: &InstanceInfoOf<T>) -> DispatchResult {
		ensure!(nft.transferable, Error::<T>::NonTransferable);
		Ok(())
	}
```

Comment (bmacer) [reply pallets/rmrk-core/src/functions.rs:376]: also would have to tweak rmrk-market lib.rs, 245-249:
```
			let nft = pallet_rmrk_core::Pallet::<T>::nfts(collection_id, nft_id)
				.ok_or(Error::<T>::TokenDoesNotExist)?;

			// Check NFT is transferable
			pallet_rmrk_core::Pallet::<T>::check_is_transferable(&nft)?;
```

File: pallets/rmrk-market/src/lib.rs [hunk 2]
```diff
@@ -242,11 +242,12 @@ pub mod pallet {
...
 
-			// Check NFT is transferable
-			let is_transferable = pallet_rmrk_core::Pallet::<T>::nfts(collection_id, nft_id)
+			let transferable = pallet_rmrk_core::Pallet::<T>::nfts(collection_id, nft_id)
 				.ok_or(Error::<T>::TokenDoesNotExist)?
 				.transferable;
-			ensure!(is_transferable, Error::<T>::NonTransferable);
+
+			// Check NFT is transferable
+			pallet_rmrk_core::Pallet::<T>::check_is_transferable(transferable)?;
```
Comment (bmacer) [inline pallets/rmrk-market/src/lib.rs:250]: If adjusting my above comment, would have to tweak this too:
```
			let nft = pallet_rmrk_core::Pallet::<T>::nfts(collection_id, nft_id)
				.ok_or(Error::<T>::TokenDoesNotExist)?;

			// Check NFT is transferable
			pallet_rmrk_core::Pallet::<T>::check_is_transferable(&nft)?;
```

**Review by bmacer — APPROVED** [review]
Consider changing `check_is_transferable` to accept `&nft` but looks good otherwise or if decision is to keep if with `bool`

File: pallets/rmrk-core/src/functions.rs (hunk 1 above)
Comment (ilionic) [reply pallets/rmrk-core/src/functions.rs:376]: Agree, thanks for the tip.

**Comment by HashWarlock** [comment]
Looks good. I think we should add this as a disclaimer or a warning for any parachain looking to integrate the RMRK pallet into their runtime that they should disable Uniques Calls to avoid the scenario below. 

In the future, when a RMRK NFT is teleported to another chain with Uniques Calls enabled there could be a situation where a non-transferable could be sent to another account. This could be an issue in the future if that scenario were to happen.

**Comment by ilionic** [comment]
Thank you, made a new issue to address such disclaimer / warning.

**Pull request merged**